### PR TITLE
feat(*): run pre-up, post-deploy, and cleanup tasks in draft

### PR DESCRIPTION
**Pull request by michelleN**
See #370 for context

### CODE REVIEW

File: examples/example-go/.draft/tasks.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+dependencies:
+  #- helm install stable/mysql --set mysqlUser=root,mysqlAllowEmptyPassword=true,mysqlDatabase=tzone_development --name zeroed-bison
+setup:
+  - kubectl exec $POD_NAME -- bash -c 'cd /app && RAILS_ENV=development bin/rake db:migrate'
```
Comment (radu-matei) [inline examples/example-go/.draft/tasks.yaml]: Is this task supposed to be in `example-go`?

File: cmd/draft/connect.go [hunk 2]
```diff
@@ -102,6 +103,10 @@ func (cn *connectCmd) run(runningEnvironment string) (err error) {
 		}
 	}
 
+	go func() {
```
Comment (bacongobbler) [inline cmd/draft/connect.go]: Can we simplify this down to just `go runTasks(tasks.Setup, connection.PodName)` or am I missing something?

File: cmd/draft/delete.go [hunk 3]
```diff
@@ -93,5 +94,10 @@ func Delete(app string) error {
 	if _, err := store.DeleteBuilds(context.Background(), app); err != nil {
 		return err
 	}
+
+	taskList, _ := tasks.Load(tasksYAMLFile)
```
Comment (bacongobbler) [inline cmd/draft/delete.go]: would it be useful to display the error in a debug statement?

File: cmd/draft/up.go [hunk 4]
```diff
@@ -182,3 +186,32 @@ func (u *upCmd) run(environment string) (err error) {
 
 	return nil
 }
+
+func runTasks(kind, podName string) error {
+	taskList, err := tasks.Load(tasksYAMLFile)
+	if err != nil && err != tasks.ErrNoTaskFile {
```
Comment (bacongobbler) [inline cmd/draft/up.go]: Would we want this to be an OR instead? I think the logic we want is if the error is non-nil OR the task file does not exist.

Also, if the task file does not exist, wouldn't we want to return an error?

File: pkg/draft/local/local.go [hunk 5]
```diff
@@ -102,6 +103,10 @@ func (a *App) Connect(clientset kubernetes.Interface, clientConfig *restclient.C
 			})
 		}
 
+		//TODO: TAKE OUT THIS IS SO BAD FOR LOTS OF REASONS
```
Comment (bacongobbler) [inline pkg/draft/local/local.go]: I'm guessing this was just for debugging purposes? :)

File: pkg/tasks/tasks.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		return nil, err
+	}
+
+	t := Tasks{}
+	tFile, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading task file: %v:%v", path, err)
+	}
+
+	err = yaml.Unmarshal(tFile, &t)
```
Comment (bacongobbler) [inline pkg/tasks/tasks.go]: I know it's just a PoC, but do we want to make this a TOML file instead so it matches draft.toml?

**Comment by squillace** [comment]
Ralph likes to speak of himself in the third person and also he mentioned toml to michelle as well. Toml's the bomb.

File: cmd/draft/up.go (hunk 4 above)
Comment (bacongobbler) [reply cmd/draft/up.go]: oh wait never mind... Let me think through this myself.

err == nil: (false && true) = false
err == SomeOtherErr: (true && true) = true
err == tasks.ErrNoTaskFile: (true && false) = false

You're right. Branch logic is hard sometimes.

**Comment by michelleN** [comment]
Update: Changing all the yamls to toml by popular demand

**Comment by bacongobbler** [comment]
I'm a fan of the latter as well, though I don't have much skin in the game at this point knowing the implications behind either implementation. :)

**Comment by michelleN** [comment]
Still working on this but it's usable now! needs docs and an example and more refactoring tho. _coming soon_

**Comment by squillace** [comment]
@michelleN you get that in, I'll help with docs. :-)

**Comment by bacongobbler** [comment]
When you have a working PoC after the feedback from earlier today, I'd love to give this a spin :)

**Comment by michelleN** [comment]
this is ready for another round of reviews if anyone has some spare 👀 @bacongobbler @radu-matei @fibonacci1729

File: cmd/draft/delete.go [hunk 7]
```diff
@@ -99,5 +100,17 @@ func Delete(app string) error {
 		return errors.New(grpc.ErrorDesc(err))
 	}
 
+	taskList, err := tasks.Load(tasksTOMLFile)
+	if err != nil {
+		if err == tasks.ErrNoTaskFile {
+			debug(err.Error())
```
Comment (bacongobbler) [inline cmd/draft/delete.go]: We probably want to also return nil here, along with all the other task loading calls below. From what I'm reading here, when no task file exists, it'll show the error if `--debug` is displayed then attempt to call `.Run` on a `nil` object.

File: docs/reference/dep-008.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+# Specification
+Tasks live in `.draft/tasks.toml` in an application's root directory. If a tasks.toml file is specified in a draft pack, running the `draft create` command will copy `tasks.toml` from the draft pack to `.draft/tasks.toml` in the application's root directory.
+
+The format of a tasks file is as follows:
+
+In `APP_ROOT/.draft/tasks.toml`:
+```toml
+[pre-up]
+hello = "echo hello world"
```
Comment (bacongobbler) [inline docs/reference/dep-008.md]: Can we clarify what the name of the task represents? Is it just visual or does it serve a purpose?

Also, is there a deterministic order for tasks to be run? e.g. if I define 2 pre-up tasks "hello" and "goodbye", which would be run first?

File: pkg/draft/pack/pack.go [hunk 9]
```diff
@@ -34,6 +34,11 @@ const (
 	// ChartsDir is the directory name for the packaged chart.
 	// This also doubles as the directory name for chart dependencies.
 	ChartsDir = "charts"
+	//TasksFileName is the name of the tasks file in a draft pack
+	TasksFileName = "tasks.toml"
+	//TargetTasksFileName is the name of the file where the tasks file from the
+	//  draft pack will be copied to
+	TargetTasksFileName = ".draft-tasks.toml"
```
Comment (bacongobbler) [inline pkg/draft/pack/pack.go]: We probably want to reference this in DEP 003, pointing to DEP 008 for further information on Draft Tasks

File: docs/reference/dep-008.md (hunk 8 above)
Comment (bacongobbler) [reply docs/reference/dep-008.md]: No that's perfect. Thank you! :)

File: cmd/draft/up.go [hunk 10]
```diff
@@ -170,15 +174,58 @@ func (u *upCmd) run(environment string) (err error) {
 		return fmt.Errorf("Could not get a helm client: %s", err)
 	}
 
+	taskList, err := tasks.Load(tasksTOMLFile)
+	if err != nil {
+		if err == tasks.ErrNoTaskFile {
+			debug(err.Error())
```
Comment (bacongobbler) [inline cmd/draft/up.go]: can you make this return nil as well? :)

File: pkg/draft/pack/pack.go (hunk 9 above)
Comment (michelleN) [reply pkg/draft/pack/pack.go]: done :)

**Comment by bacongobbler** [comment]
Just needs a rebase then merge away!

**Pull request merged**